src/coreclr/System.Private.CoreLib/src/System/Runtime/ControlledExecution.CoreCLR.cs

```diff
@@ -0,0 +1,177 @@
+        /// <exception cref="System.InvalidOperationException">
+        /// The current thread is already running the <see cref="ControlledExecution.Run"/> method.
+        /// </exception>
+        /// <exception cref="System.OperationCanceledException">The execution was aborted.</exception>
+        /// <remarks>
+        /// <see cref="ControlledExecution"/> enables aborting arbitrary code in a non-cooperative manner.
+        /// Doing so may corrupt the process.  This method is not recommended for use in production code
+        /// in which reliability is important.
+        /// </remarks>
+        [Obsolete(Obsoletions.ControlledExecutionRunMessage, DiagnosticId = Obsoletions.ControlledExecutionRunDiagId, UrlFormat = Obsoletions.SharedUrlFormat)]
+        public static void Run(Action action, CancellationToken cancellationToken)
+        {
+            ArgumentNullException.ThrowIfNull(action);
+
+            // Recursive ControlledExecution.Run calls are not supported
+            if (t_executing)
+                throw new InvalidOperationException(SR.InvalidOperation_NestedControlledExecutionRun);
+
+            new Execution(action, cancellationToken).Run();
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "ThreadNative_Abort")]
+        private static partial void AbortThread(ThreadHandle thread);
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern bool ResetAbortThread();
+
+        private sealed partial class Execution
+        {
+            // The state transition diagram (F means the Finished flag and so on):
+            // N  ⟶ F
+            // ↓
+            // AR ⟶ FAR
+            // ↓      ↓
+            // A  ⟶ FA
+            private enum State : int
+            {
+                None = 0,
+                Finished = 1,
+                AbortRequested = 2,
+                RunningAbort = 4
+            }
+
+            // Interpreted as a value of the State enumeration type
+            private int _state;
+            private readonly Action _action;
+            private readonly CancellationToken _cancellationToken;
+            private Thread? _thread;
+
+            public Execution(Action action, CancellationToken cancellationToken)
+            {
+                _action = action;
+                _cancellationToken = cancellationToken;
+            }
+
+            public void Run()
+            {
+                Debug.Assert(_state == (int)State.None && _thread == null);
+                _thread = Thread.CurrentThread;
+
+                try
+                {
+                    try
+                    {
+                        t_executing = true;
+                        // Cannot Dispose this registration in a finally or a catch block as that may deadlock with AbortThread
+                        _cancellationToken.UnsafeRegister(e => ((Execution)e!).Abort(), this);
```

Comment:
Even if you don't take the overall changes I suggested (but I hope you do :), at a minimum the CancellationTokenRegistration returned from UnsafeRegister must be Dispose'd or Unregister'd when it's no longer wanted.  Otherwise, it'll leak the registration into the CancellationToken.